### PR TITLE
issue3045/bug plate creator uses number from non sanger barcodes to generate child barcodes

### LINKED ISSUE #3045: GPL-850 BUG - Plate creator uses number from non-sanger barcodes to generate child barcodes
**Describe the bug**
When generating plates via a Plate::Creator, sequencescape will attempt to re-use the existing barcode number in the event that the prefix changes. Eg. DN12345 makes WD12345.

This behaviour should *not* trigger for foreign barcode formats, as it increases the risk of barcodes clashing with other teams. In practice our switch to code39 format barcodes, rather than the ean13 encoded forms, makes actual clashes unlikely.

**RT Ticket Number**
If applicable

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to page '...'
2. Click on button '...'
3. See error

**Expected behaviour**
What was expected to happen.

**Screenshots**

**Desktop (please complete the following information):**
* Windows or Mac?
* Browser Chrome, Firefox, Safari or other?
* Browser version (use 'About' to look up)?

**Additional context**

### PR DESCRIPTION
**Pull request by JamesGlover**
Prevent reuse of barcode number of external barcodes

Fixes #3045

Split into two commits, the first is the fix for the issue,
the second a refactor pass to fix the rubocop todos, and some
simple stuff I spotted tht the time.

This implements its own plate creation and transfer behaviour.
We should probably unify it with the systems used by Limber
etc, or as a bare minimum use transfer requests. However, I've
avoided that here as:

1) Risk.
2) It actually changes the behaviour slightly. A stamp of a partial cherrypick
   with this method results in a partial child, which could potentially be
   picked onto further. (As it only dups existing wells) The 'proper' approach
   would result in a plate with 96 wells, but some of the empty.

The behaviour in point 2 is definitely 'surprising,' but is may be a reason to
update the cherrypicking behaviour before this.

- Exclude foreign barcodes from child plate barcode generation
- Make rubocop happy and slight refactor

### CODE REVIEW

**Comment by emrojo** [comment]
I'm trying to understand the logic behind the plate creation and barcodes number. Could you check if this I write is correct?

* You can have a plate with different barcodes
* There are two types of barcodes you can have: Sanger barcodes and Foreign barcodes
* Sanger barcodes are barcodes with prefix DN, WD, ... (original SS plates)
* Foreign barcodes can have a lot of other different formats
* Both barcode types have in common that they have a prefix and a number
* There are 3 situations with a plate that has a barcode:
  * A - You can have a maximum of 2 barcodes, 1 Sanger barcode and 1 Foreign barcode
  * B - Or you can have a single barcode 1 Sanger barcode
  * C - Or you can have a single barcode 1 Foreign barcode
* When you create a new plate from a parent plate:
   1. When the parent plate has 1 Sanger barcode and 1 Foreign barcode (Type A), then the child plate will use the number of the Sanger barcode to try to create a new barcode with the prefix for its plate purpose; if the same prefix and number exists in the database, it will create a new number, otherwise, it will reuse the same number (but use the new different prefix).
   2. When the parent plate has only 1 Sanger barcode(Type B), it is same situation as Type A before
   3. When the parent plate has only 1 Foreign barcode(Type C), it will always generate a new number for Sanger barcode

**Review by emrojo — APPROVED** [review]
Only linting suggestions, but all ok. I wrote a comment about how I think the logic behind creating plates from different barcodes works. Could you check if that is correct?

**Comment by JamesGlover** [comment]
> I'm trying to understand the logic behind the plate creation and barcodes number. Could you check if this I write is correct?

Yep, you've got it pretty much correct

> * Both barcode types have in common that they have a prefix and a number

Strictly speaking this isn't a requirement of foreign barcodes. For example, the fluidigm barcode format doesn't have a prefix.

>   * A - You can have a maximum of 2 barcodes, 1 Sanger barcode and 1 Foreign barcode

We actually allow an unlimited number of foreign barcodes, although I don't think we see more than one in practice.

>   * C - Or you can have a single barcode 1 Foreign barcode

In practice I'm not sure how many of these we have kicking around, I think most routes in end up generating Sanger barcodes as well. Its definitely supported though.

File: app/models/plate/creator.rb [hunk 1]
```diff
@@ -184,18 +193,15 @@ def create_plates(source_plate_barcodes, current_user, creator_parameters = nil)
...
-          Plate.with_barcode(scanned).eager_load(wells: :aliquots).first or
-          fail_with_error("Could not find plate with machine barcode #{scanned.inspect}")
+      scanned_barcodes.flat_map do |scanned|
+        plate = Plate.with_barcode(scanned).eager_load(wells: :aliquots).find_by_barcode(scanned) ||
+                fail_with_error("Could not find plate with machine barcode #{scanned.inspect}")
 
-        unless can_create_plates?(plate)
-          fail_with_error("Scanned plate #{scanned} has a purpose #{plate.purpose.name} not valid for creating [#{plate_purposes.map(&:name).join(',')}]")
-        end
+        fail_with_error("Scanned plate #{scanned} has a purpose #{plate.purpose.name} not valid for creating [#{plate_purposes.map(&:name).join(',')}]") unless can_create_plates?(plate)
```
[the remark this answers is not in the document]
Comment (JamesGlover) [reply app/models/plate/creator.rb:200]: Ugh, yes, I agree. I actually prefer the former in this case, but that was a rubocop auto-correct. Should be able to tidy this one up though.

File: app/models/plate/creator.rb [hunk 2]
```diff
@@ -184,18 +193,15 @@ def create_plates(source_plate_barcodes, current_user, creator_parameters = nil)
...
 
-        destinations = create_child_plates_from(plate, current_user, creator_parameters)
-        add_created_plates(plate, destinations)
-        plates.concat(destinations)
+        create_child_plates_from(plate, current_user, creator_parameters).tap do |destinations|
+          add_created_plates(plate, destinations)
+        end
```
[the remark this answers is not in the document]
Comment (JamesGlover) [reply app/models/plate/creator.rb:204]: Before it was using each_with_object I switched to to using flat_map instead.

```ruby
[[:a],:b,[[:c]]].flat_map {|a| a } #=> [:a, :b, [:c]]
```

So flat_map is a bit like map, but will concatenate on an array if it receives one. You sometimes see people claiming it is the same a .map {}.flatten, but if you look at the example above, it only goes one level deep.

Its a really nice method for joining together and aggregating collections. It wasn't available when the code was first written, although I have a horrible feeling it was around when I made the switch to each_with_object.

**Pull request merged**